week9/answers.sql

```diff
@@ -1,3 +1,165 @@
+    CONSTRAINT  friends_user_fk_users
+    FOREIGN KEY (user_id)
+    REFERENCES users (user_id)
+    ON UPDATE CASCADE
+    ON DELETE CASCADE,
+    CONSTRAINT  friends_friend_fk_users
+    FOREIGN KEY (friend_id)
+    REFERENCES users (user_id)
+    ON UPDATE CASCADE
+    ON DELETE CASCADE
+);
+
+-- Create 'posts' table
+CREATE TABLE posts (
+    post_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    user_id INT UNSIGNED NOT NULL,
+    created_on TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+    updated_on TIMESTAMP DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
+    content VARCHAR(255) NOT NULL,
+    CONSTRAINT  posts_fk_users
+    FOREIGN KEY (user_id)
+    REFERENCES users (user_id)
+    ON UPDATE CASCADE
+    ON DELETE CASCADE
+);
+
+-- Create 'notifications' table
+CREATE TABLE notifications (
+    notification_id INT UNSIGNED PRIMARY KEY NOT NULL AUTO_INCREMENT,
+    user_id INT UNSIGNED NOT NULL,
+    post_id INT UNSIGNED NOT NULL,
+    CONSTRAINT  notifications_fk_users
+    FOREIGN KEY (user_id)
+    REFERENCES users (user_id)
+    ON UPDATE CASCADE
+    ON DELETE CASCADE,
+    CONSTRAINT  notifications_fk_posts
+    FOREIGN KEY (post_id)
+    REFERENCES posts (post_id)
+    ON UPDATE CASCADE
+    ON DELETE CASCADE
+);
+
+
+
+-- Create notification_posts VIEW
+CREATE VIEW notification_posts AS
+SELECT n.user_id, u.first_name, u.last_name, p.post_id, p.content
+FROM notifications n
+LEFT OUTER JOIN posts p ON n.post_id = p.post_id
+LEFT OUTER JOIN users u ON p.user_id = u.user_id;
+
+
+-- This uses a CURSOR
+-- Here's some psuedo-code:
+-- Create a new post with the desired message.
+-- For each user that isn't the newly joined user:
+-- Add a notification with that users's ID and the post ID.
+DELIMITER ;;
+CREATE TRIGGER after_user_insert
+AFTER INSERT ON users
+FOR EACH ROW
+BEGIN
+    DECLARE new_user_id INT UNSIGNED;
+    DECLARE new_post_id INT UNSIGNED;
+    DECLARE new_friend_id INT UNSIGNED;
+    DECLARE row_not_found TINYINT DEFAULT FALSE;
+	
+    DECLARE users_cursor CURSOR FOR 
+		SELECT u.user_id 
+		FROM users u
+		WHERE u.user_id != NEW.user_id;
+            
+    DECLARE CONTINUE HANDLER FOR NOT FOUND
+    SET row_not_found = TRUE;
+        
+    INSERT INTO posts (posts.user_id, posts.content) VALUES (NEW.user_id, concat(NEW.first_name,  ' ' , NEW.last_name, ' ' , 'just joined!'));
+    SET new_post_id = LAST_INSERT_ID();
+    
+    OPEN users_cursor;
+    user_loop: LOOP
+    FETCH users_cursor INTO new_user_id;
+    IF row_not_found THEN
+	LEAVE user_loop;
+    END IF;
+    INSERT INTO notifications (user_id, post_id) 
+    VALUES 
+    (new_user_id, new_post_id);
+  END LOOP;
+  CLOSE users_cursor; -- LOOK I CLOSED MY CURSOR THIS TIME!!! :D
+END;;
+DELIMITER ;
+
+
+CREATE EVENT IF NOT EXISTS remove_old_sessions
+ON SCHEDULE EVERY 10 SECOND
+DO
+DELETE FROM sessions 
+WHERE updated_on < NOW() - INTERVAL 2 HOUR;
+
+
+DELIMITER ;;
+CREATE PROCEDURE add_post(IN user_id INT, IN content TEXT)
```

Comment:
Because the content field in posts is a VARCHAR(255), I highly recommend using the same data type, here. Same with user_id, which is UNSIGNED in the table declarations. It would be more efficient for those to be caught here rather than after running the following code. (If someone tried to pass -1 as a user_id, the PROCEDURE would never run if the datatype matched INT UNSIGNED, but with an INT, the code will all the way to "INSERT INTO posts".)